### PR TITLE
4.x: PrivateLink Support Phase 1: Configuration ⚙️

### PR DESCRIPTION
**Pull request by nikagra**
# 4.x: PrivateLink Support Phase 1: Configuration ⚙️

**Note:** Depends on [#java-native-protocol/45](java-native-protocol/pull/45) :heavy_check_mark: 
## Overview

Implements the configuration API for Client Routes feature to support PrivateLink-style cloud deployments (ScyllaDB Cloud). This is Phase 1 of the implementation - provides programmatic configuration without the full handler implementation.

**Jira:** [DRIVER-88](https://scylladb.atlassian.net/browse/DRIVER-88)

### New API Classes

- **`ClientRoutesEndpoint`** - Represents a single endpoint with connection ID and optional DNS address
- **`ClientRoutesConfig`** - Immutable configuration container with builder pattern
- DNS cache with configurable expiry

### SessionBuilder Integration

- Added `withClientRoutesConfig(ClientRoutesConfig)` method
- Validates mutual exclusivity with custom `AddressTranslator`
- Automatically derives seed hosts from endpoint addresses when no contact points provided
- Supports IPv4, IPv6, and hostname formats with URI-based parsing

### Configuration

- Added `advanced.client-routes.table-name` option (default: `system.client_routes`)
- Extended `ProgrammaticArguments` to pass configuration through session initialization
- Updated `reference.conf` with inline documentation

### Address Parsing

- URI-based implementation (RFC 3986 compliant)
- Handles IPv4, IPv6 (bracketed), and hostnames
- Port validation (1-65535) with helpful error messages
- Default port: 9042

## Supported Address Formats

```java
// IPv4
"192.168.1.1:9042"
"192.168.1.1"         // defaults to 9042

// IPv6 (must use brackets)
"[::1]:9042"
"[2001:db8::1]:9042"
"[::1]"               // defaults to 9042

// Hostname
"my-cluster.scylladb.com:9042"
"my-cluster.scylladb.com"
```

## Usage Example

```java
ClientRoutesConfig config = ClientRoutesConfig.builder()
    .addEndpoint(new ClientRoutesEndpoint(
        UUID.fromString("12345678-1234-1234-1234-123456789012"),
        "my-cluster.us-east-1.aws.scylladb.com:9042"))
    .build();

CqlSession session = CqlSession.builder()
    .withClientRoutesConfig(config)
    .withLocalDatacenter("datacenter1")
    .build();
```

## Testing

- **10 unit tests** - All passing
  - `ClientRoutesConfigTest` (8 tests)
  - `ClientRoutesSessionBuilderTest` (2 tests, refactored to use Mockito)
- **8 validation tests** - Address parsing edge cases
  - `ClientRoutesValidationTest`
- Test fixes for `TypedDriverOptionTest` and `MapBasedDriverConfigLoaderTest`

## Documentation

- Updated `manual/core/address_resolution/README.md` with Client Routes section
- Updated `reference.conf` with configuration options
- Updated `changelog/README.md` for version 4.19.0

## Validation

✅ All unit tests passing (18 tests total)  
✅ No compilation errors  
✅ Mutual exclusivity validation working  
✅ Seed host derivation working  
✅ IPv6 support working  
✅ Error messages clear and helpful  

## What's NOT Included (Future Work)

This PR provides the **configuration API only**. The following are intentionally deferred to future PRs:

- ❌ Client Routes Handler implementation (DNS resolution, caching, address translation)
- ❌ Address Translator V2 API (Host ID-based translation)
- ❌ `CLIENT_ROUTES_CHANGE` event handling (stub exists in `ControlConnection`)
- ❌ Control connection event registration
- ❌ Integration tests with mock cluster

## Design Decisions

1. **Programmatic-only configuration** - Connection IDs are sensitive; no file-based config
2. **Mutual exclusivity with AddressTranslator** - Prevents conflicts
3. **URI-based parsing** - Standards-compliant, handles IPv6 correctly, simple implementation
4. **Immutable configuration** - Thread-safe, prevents accidental modifications
5. **Mockito for tests** - Eliminated 92 lines of boilerplate code

## Breaking Changes

None - this is a purely additive change.

## Migration Guide

N/A - New feature, no migration needed.

## Review Checklist

- [ ] API design review (immutability, builder pattern, validation)
- [ ] SessionBuilder integration review
- [ ] URI-based address parsing correctness
- [ ] Test coverage adequate
- [ ] Documentation clarity
- [ ] Error messages helpful
- [ ] No regressions in existing tests

## Links

- Server PR: [scylladb/scylladb#27323](https://github.com/scylladb/scylladb/pull/27323)
- Go reference: [gocql/gocql#675](https://github.com/gocql/gocql/pull/675)
- Design: [Core Design Doc](https://scylladb.atlassian.net/wiki/spaces/RND/pages/163053946/Private+Link+in+Core)

---

Related to [DRIVER-86](https://scylladb.atlassian.net/browse/DRIVER-86), [DRIVER-88](https://scylladb.atlassian.net/browse/DRIVER-88)

[DRIVER-88]: https://scylladb.atlassian.net/browse/DRIVER-88?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DRIVER-86]: https://scylladb.atlassian.net/browse/DRIVER-86?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/src/main/java/com/datastax/oss/driver/api/core/session/SessionBuilder.java [hunk 1]
```diff
@@ -898,6 +941,44 @@ protected final CompletionStage<CqlSession> buildDefaultSessionAsync() {
...
+                InetSocketAddress socketAddress =
+                    AddressParser.parseContactPoint(addr, endpoint.getConnectionId());
+                programmaticContactPoints.add(new DefaultEndPoint(socketAddress));
+
+              } catch (Exception e) {
+                throw new IllegalArgumentException(
+                    String.format(
+                        "Failed to parse client routes endpoint address '%s' (connection ID: %s): %s",
+                        addr, endpoint.getConnectionId(), e.getMessage()),
+                    e);
```
[the remark this answers is not in the document]
Comment (nikagra) [reply core/src/main/java/com/datastax/oss/driver/api/core/session/SessionBuilder.java]: Comment makes sense, code suggestion does not. Exception from `AddressParser` will be propagated naturally, no extra catching and re-throwing required

File: core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesConfig.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+     * name from the configuration ({@code system.client_routes}).
+     *
+     * @param tableName the table name to use.
+     * @return this builder.
+     */
+    @NonNull
+    public Builder withTableName(@Nullable String tableName) {
+      this.tableName = tableName;
+      return this;
+    }
```
Comment (dkropachev) [inline core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesConfig.java:206]: please mark it for tests only

File: core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+/**
+ * Represents a client routes endpoint for PrivateLink-style deployments.
+ *
+ * <p>Each endpoint corresponds to a connection ID in the {@code system.client_routes} table, with
+ * an optional connection address that can be used as a seed host for initial connection.
+ */
+@Immutable
+public class ClientRoutesEndpoint {
+
+  private final UUID connectionId;
```
Comment (dkropachev) [inline core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesEndpoint.java:35]: as far as i remember connection id is not a UUID, it is a string

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+/**
+ * Utility class for parsing network addresses. This is used internally for parsing contact points
+ * and client routes endpoints.
+ *
+ * <p>This class is part of the internal API and is not intended for public use. Classes under
+ * {@code com.datastax.oss.driver.internal.*} are not user-facing and may change without notice.
+ */
+public class AddressParser {
+
+  private static final int DEFAULT_PORT = 9042;
```
Comment (dkropachev) [inline core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java]: you need to pull this from the config

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    try {
+      // Add scheme to make it a valid URI for parsing
+      // URI class handles IPv6 brackets, hostname, and port correctly
+      String uriString = address.contains("://") ? address : "cql://" + address;
+      java.net.URI uri = new java.net.URI(uriString);
+
+      String host = uri.getHost();
+      int port = uri.getPort();
```
Comment (dkropachev) [inline core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java]: please drop it

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+   *   <li>[ipv6]:port (e.g., "[::1]:9042", "[2001:db8::1]:9042")
+   *   <li>[ipv6] (defaults to port 9042)
+   * </ul>
+   *
+   * @param address the address string to parse (must not be null)
+   * @param connectionId the connection ID for error messages (can be null for generic parsing)
+   * @return an InetSocketAddress
+   * @throws IllegalArgumentException if the address is null, empty, or has an invalid format
+   */
+  public static InetSocketAddress parseContactPoint(String address, UUID connectionId) {
```
Comment (dkropachev) [inline core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java]: move this code `SessionBuilder`

File: core/src/main/resources/reference.conf [hunk 7]
```diff
@@ -1110,6 +1110,31 @@ datastax-java-driver {
     # advertised-hostname = mycustomhostname
   }
 
+  # Client routes configuration for PrivateLink-style deployments.
```
Comment (dkropachev) [inline core/src/main/resources/reference.conf]: Make it cloud agnostic, there is private link on aws, private service connect on gcp, probably something else on azure, if you worked with azure, you have no idea what private link is.
Describe it from the stand point of the way nodes are exposed instead.

File: manual/core/address_resolution/README.md [hunk 8]
```diff
@@ -118,6 +118,52 @@ datastax-java-driver.advanced.address-translator.class = com.mycompany.MyAddress
 Note: the contact points provided while creating the `CqlSession` are not translated, only addresses
 retrieved from or sent by Cassandra nodes are.
 
+### Client Routes (PrivateLink deployments)
+
+For cloud deployments using PrivateLink or similar private endpoint technologies (such as ScyllaDB Cloud), nodes are
+accessed through private DNS endpoints rather than direct IP addresses. The driver provides a client routes feature
+to handle this topology.
```
Comment (dkropachev) [inline manual/core/address_resolution/README.md]: make it cloud agnosting.

File: core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesEndpoint.java (hunk 3 above)
Comment (nikagra) [reply core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesEndpoint.java:35]: Based on scylla-drivers#115:
```cql
CREATE TABLE system.client_routes (
    connection_id uuid,
    host_id uuid,
    address text,          -- DNS hostname of the PrivateLink endpoint
    port int,              -- CQL plaintext port
    tls_port int,          -- CQL TLS port
    alternator_port int,
    alternator_https_port int,
    datacenter text,
    rack text,
    PRIMARY KEY (connection_id, host_id)
);
```
it is `uuid`

File: core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesConfig.java (hunk 2 above)
Comment (nikagra) [reply core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesConfig.java:206]: Resolved in phase 2 PR #804

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java (hunk 5 above)
Comment (nikagra) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java]: Please elaborate. Do you mean I should not get port from URI?

Comment (dkropachev) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java]: i mean there should be no uri, no other drivers support scheme like this one, it is a regular endpoint string in all of em

File: core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesEndpoint.java (hunk 3 above)
Comment (dkropachev) [reply core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesEndpoint.java:35]: i believe it is outdated, check it on scylla

File: core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java (hunk 4 above)
Comment (nikagra) [reply core/src/main/java/com/datastax/oss/driver/internal/core/util/AddressParser.java]: Please correct me if I'm wrong, but there is no `defaultPort` option in the driver. Basic contact endpoints are provided with ports as below:
```conf
datastax-java-driver {
  basic.contact-points = [ "127.0.0.1:9042", "127.0.0.2:9042" ]
```
If client routes endpoints are provided without a port, then a separate config may be needed.

File: core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesEndpoint.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  public ClientRoutesEndpoint(@NonNull UUID connectionId, @Nullable String connectionAddr) {
+    this.connectionId = Objects.requireNonNull(connectionId, "connectionId must not be null");
+    this.connectionAddr = connectionAddr;
+  }
+
+  /** Returns the connection ID for this endpoint. */
+  @NonNull
+  public UUID getConnectionId() {
+    return connectionId;
+  }
```
Comment (dkropachev) [inline core/src/main/java/com/datastax/oss/driver/api/core/config/ClientRoutesEndpoint.java:62]: It is not UUID

**Comment by nikagra** [comment]
See #807
